securedrop_client/export/printer.py

```diff
@@ -0,0 +1,275 @@
+from typing import Callable, NewType
+
+from PyQt5.QtCore import QObject, QState, QStateMachine, QTimer, pyqtSignal, pyqtSlot
+
+from .cli import Error as CLIError
+from .service import Service
+
+DEFAULT_POLLING_INTERVAL_IN_MILLISECONDS = 2000
+
+
+class Printer(QObject):
+    """Allows to enqueue printing jobs and track the status of the printing queue.
+
+    It can be treated like a printer, but technically speaking it only interfaces with
+    a printing queue, not an actual printer. Success or failure enqueing jobs can be tracked,
+    and is the best proxy we currently have to tracking the outcome of printing operations."""
+
+    status_changed = pyqtSignal()
+    job_done = pyqtSignal(str)
+    job_failed = pyqtSignal(str, str)
+
+    client_connected = pyqtSignal()
+    last_client_disconnected = pyqtSignal()
+
+    Status = NewType("Status", str)
+    StatusUnknown = Status("unknown-sf5fd")
+    StatusReady = Status("ready-83jf3")
+    StatusUnreachable = Status("unreachable-120a0")
+
+    def __init__(
+        self,
+        printing_service: Service,
+        polling_interval_in_milliseconds: int = DEFAULT_POLLING_INTERVAL_IN_MILLISECONDS,
+    ) -> None:
+        super().__init__()
+
+        self._connected_clients = 0
+        self._printing_service = printing_service
+        self._poller = _Poller(polling_interval_in_milliseconds)
+        self._cache = _StatusCache(self._printing_service)
+
+        # Accept that the status is unknown if we don't watch the printer for a bit.
+        self._cache.clear_on(self._poller.paused.entered)
+        self._cache.on_change_emit(self.status_changed)
+
+        # This is a blocking call, which is no good.
+        # self._poller.poll_by(lambda: self._printing_service.check_printer())
+        # Alternatively, by taking advantage of the printing service features to loosen coupling:
+        self._printing_service.connect_signals(printer_check_requested=self._poller.polling.entered)
```

Comment:
Note @rocodes I think this will go unaddressed for this round, unless you point it out too. (If two of us have feelings on it, I'd fix it, but I would sit on my own feelings if it's just me.)